// @mondaycom/apps-cli/dist/services/share/deploy.js
import { Listr } from 'listr2';
import { buildAssetToDeployTask, buildClientZip, deployClientZip, handleDeploymentTask, prepareEnvironmentTask, uploadAssetTask, } from '../push-service.js';
export const getTasksForClientSide = (appVersionId, directoryPath) => {
    return new Listr([
        { title: 'Build asset to deploy', task: buildClientZip },
        {
            title: 'Deploying client side files',
            task: deployClientZip,
        },
    ], {
        ctx: { appVersionId, directoryPath },
    });
};
export const getTasksForServerSide = (appVersionId, directoryPath) => {
    return new Listr([
        { title: 'Build asset to deploy', task: buildAssetToDeployTask },
        {
            title: 'Preparing environment',
            task: prepareEnvironmentTask,
            enabled: ctx => Boolean(ctx.showPrepareEnvironmentTask),
        },
        {
            title: 'Uploading built asset',
            task: uploadAssetTask,
            enabled: ctx => Boolean(ctx.showUploadAssetTask),
        },
        {
            title: 'Deployment in progress',
            task: handleDeploymentTask,
            enabled: ctx => Boolean(ctx.showHandleDeploymentTask),
        },
    ], { ctx: { appVersionId, directoryPath } });
};
