// @mondaycom/apps-cli/dist/types/services/app-features-service.js
export var BUILD_TYPES;
(function (BUILD_TYPES) {
    BUILD_TYPES["CUSTOM_URL"] = "custom_url";
    BUILD_TYPES["MONDAY_CODE"] = "monday_code";
    BUILD_TYPES["MONDAY_CODE_CDN"] = "monday_code_cdn";
})(BUILD_TYPES || (BUILD_TYPES = {}));
export var AppReleaseSingleBuildCategory;
(function (AppReleaseSingleBuildCategory) {
    AppReleaseSingleBuildCategory["MondayCode"] = "monday_code";
    AppReleaseSingleBuildCategory["view"] = "view";
})(AppReleaseSingleBuildCategory || (AppReleaseSingleBuildCategory = {}));
export var AppFeatureType;
(function (AppFeatureType) {
    AppFeatureType["AppFeatureOauth"] = "AppFeatureOauth";
    AppFeatureType["AppFeatureBoardView"] = "AppFeatureBoardView";
    AppFeatureType["AppFeatureIntegration"] = "AppFeatureIntegration";
    AppFeatureType["AppFeatureSolution"] = "AppFeatureSolution";
    AppFeatureType["AppFeatureItemView"] = "AppFeatureItemView";
    AppFeatureType["AppFeatureDashboardWidget"] = "AppFeatureDashboardWidget";
    AppFeatureType["AppFeatureAccountSettingsView"] = "AppFeatureAccountSettingsView";
    AppFeatureType["AppFeatureDocActions"] = "AppFeatureDocActions";
    AppFeatureType["AppFeatureObject"] = "AppFeatureObject";
    AppFeatureType["AppFeatureWorkspaceView"] = "AppFeatureWorkspaceView";
    AppFeatureType["AppFeatureAI"] = "AppFeatureAI";
    AppFeatureType["AppFeatureAiBoardMainMenuHeader"] = "AppFeatureAiBoardMainMenuHeader";
    AppFeatureType["AppFeatureAiItemUpdateActions"] = "AppFeatureAiItemUpdateActions";
    AppFeatureType["AppFeatureAiDocSlashCommand"] = "AppFeatureAiDocSlashCommand";
    AppFeatureType["AppFeatureAiDocContextualMenu"] = "AppFeatureAiDocContextualMenu";
    AppFeatureType["AppFeatureAiDocQuickStart"] = "AppFeatureAiDocQuickStart";
    AppFeatureType["AppFeatureAiDocTopBar"] = "AppFeatureAiDocTopBar";
    AppFeatureType["AppFeatureColumnTemplate"] = "AppFeatureColumnTemplate";
    AppFeatureType["AppFeatureAiIcAssistantHelpCenter"] = "AppFeatureAiIcAssistantHelpCenter";
    AppFeatureType["AppFeatureAppWizard"] = "AppFeatureAppWizard";
    AppFeatureType["AppFeatureGroupMenuAction"] = "AppFeatureGroupMenuAction";
    AppFeatureType["AppFeatureItemMenuAction"] = "AppFeatureItemMenuAction";
    AppFeatureType["AppFeatureNotificationKind"] = "AppFeatureNotificationKind";
    AppFeatureType["AppFeatureBlock"] = "AppFeatureBlock";
    AppFeatureType["AppFeatureItemBatchAction"] = "AppFeatureItemBatchAction";
    AppFeatureType["AppFeatureAiFormula"] = "AppFeatureAiFormula";
    AppFeatureType["AppFeatureAiItemEmailsAndActivitiesActions"] = "AppFeatureAiItemEmailsAndActivitiesActions";
    AppFeatureType["AppFeatureFieldType"] = "AppFeatureFieldType";
    AppFeatureType["AppFeatureProduct"] = "AppFeatureProduct";
})(AppFeatureType || (AppFeatureType = {}));
