// @mondaycom/apps-cli/dist/consts/urls.js
const BASE_APPS_URL = '/api/apps';
const BASE_APP_VERSIONS_URL = '/api/app-versions';
const BASE_MONDAY_CODE_URL = '/api/code';
export const appVersionIdBaseUrl = (appVersionId) => {
    return `/api/code/${appVersionId}`;
};
export const getAppVersionDeploymentStatusUrl = (appVersionId) => {
    return `${appVersionIdBaseUrl(appVersionId)}/deployments`;
};
export const getAppFeaturesUrl = (appVersionId, types) => {
    const url = `${BASE_APP_VERSIONS_URL}/${appVersionId}/app-features`;
    const appFeatureTypes = types?.map((type, index) => `type[${index}]=${type}`).join('&');
    return appFeatureTypes ? `${url}?${appFeatureTypes}` : url;
};
export const getCreateAppFeatureUrl = (appId, appVersionId) => {
    return `${BASE_APPS_URL}/${appId}/app-versions/${appVersionId}/app-features`;
};
export const getCreateAppFeatureReleaseUrl = (appId, appVersionId, appFeatureId) => {
    return `${BASE_APPS_URL}/${appId}/versions/${appVersionId}/app-features/${appFeatureId}/releases`;
};
export const getDeploymentSignedUrl = (appVersionId) => {
    return `${appVersionIdBaseUrl(appVersionId)}/deployments/signed-url`;
};
export const getDeploymentClientUpload = (appVersionId) => {
    return `${appVersionIdBaseUrl(appVersionId)}/deployments`;
};
export const getStorageItemsSearchUrl = (appId, clientAccountId, term) => {
    return `/api/storage/app/${appId}/account/${clientAccountId}/records?term=${encodeURI(term)}`;
};
export const getStorageItemsExportUrl = (appId, clientAccountId) => {
    return `/api/storage/app/${appId}/account/${clientAccountId}/records/export`;
};
export const getLogsStreamForAppVersionIdUrl = (appVersionId, logsType, logsFilterCriteria) => {
    const logsFilterCriteriaParams = logsFilterCriteria
        ? `&fromDate=${logsFilterCriteria.fromDate.getTime()}&toDate=${logsFilterCriteria.toDate.getTime()}&text=${encodeURIComponent(logsFilterCriteria.text || '')}`
        : '';
    return `${appVersionIdBaseUrl(appVersionId)}/logs?type=${logsType}${logsFilterCriteriaParams}`;
};
export const listAppsUrl = () => {
    return BASE_APPS_URL;
};
export const listAppVersionsByAppIdUrl = (appId) => {
    return `${BASE_APPS_URL}/${appId}/versions`;
};
export const appEnvironmentUrl = (appId, key) => {
    return `/api/code/${appId}/env/${key}`;
};
export const appEnvironmentKeysUrl = (appId) => {
    return `/api/code/${appId}/env-keys`;
};
export const appReleasesUrl = (appVersionId) => {
    return `/apps_ms/app-versions/${appVersionId}/releases`;
};
export const generateTunnelingTokenUrl = () => {
    return `${BASE_MONDAY_CODE_URL}/tunnel-token`;
};
