// @mondaycom/apps-cli/dist/utils/time-enum.js
const SECOND_IN_MILLISECOND = 1000;
export var TIME_IN_SECONDS;
(function (TIME_IN_SECONDS) {
    TIME_IN_SECONDS[TIME_IN_SECONDS["YEAR"] = Number(365 * 24 * 60 * 60)] = "YEAR";
    TIME_IN_SECONDS[TIME_IN_SECONDS["MONTH31Days"] = Number(31 * 24 * 60 * 60)] = "MONTH31Days";
    TIME_IN_SECONDS[TIME_IN_SECONDS["MONTH30Days"] = Number(30 * 24 * 60 * 60)] = "MONTH30Days";
    TIME_IN_SECONDS[TIME_IN_SECONDS["MONTH29Days"] = Number(29 * 24 * 60 * 60)] = "MONTH29Days";
    TIME_IN_SECONDS[TIME_IN_SECONDS["MONTH28Days"] = Number(28 * 24 * 60 * 60)] = "MONTH28Days";
    TIME_IN_SECONDS[TIME_IN_SECONDS["DAY"] = Number(24 * 60 * 60)] = "DAY";
    TIME_IN_SECONDS[TIME_IN_SECONDS["HOUR"] = Number(60 * 60)] = "HOUR";
    TIME_IN_SECONDS[TIME_IN_SECONDS["MINUTE"] = 60] = "MINUTE";
})(TIME_IN_SECONDS || (TIME_IN_SECONDS = {}));
export var TIME_IN_MILLISECONDS;
(function (TIME_IN_MILLISECONDS) {
    TIME_IN_MILLISECONDS[TIME_IN_MILLISECONDS["YEAR"] = Number(365 * 24 * 60 * 60 * SECOND_IN_MILLISECOND)] = "YEAR";
    TIME_IN_MILLISECONDS[TIME_IN_MILLISECONDS["MONTH31Days"] = Number(31 * 24 * 60 * 60 * SECOND_IN_MILLISECOND)] = "MONTH31Days";
    TIME_IN_MILLISECONDS[TIME_IN_MILLISECONDS["MONTH30Days"] = Number(30 * 24 * 60 * 60 * SECOND_IN_MILLISECOND)] = "MONTH30Days";
    TIME_IN_MILLISECONDS[TIME_IN_MILLISECONDS["MONTH29Days"] = Number(29 * 24 * 60 * 60 * SECOND_IN_MILLISECOND)] = "MONTH29Days";
    TIME_IN_MILLISECONDS[TIME_IN_MILLISECONDS["MONTH28Days"] = Number(28 * 24 * 60 * 60 * SECOND_IN_MILLISECOND)] = "MONTH28Days";
    TIME_IN_MILLISECONDS[TIME_IN_MILLISECONDS["DAY"] = Number(24 * 60 * 60 * SECOND_IN_MILLISECOND)] = "DAY";
    TIME_IN_MILLISECONDS[TIME_IN_MILLISECONDS["HOUR"] = Number(60 * 60 * SECOND_IN_MILLISECOND)] = "HOUR";
    TIME_IN_MILLISECONDS[TIME_IN_MILLISECONDS["MINUTE"] = Number(60 * SECOND_IN_MILLISECOND)] = "MINUTE";
    TIME_IN_MILLISECONDS[TIME_IN_MILLISECONDS["SECOND"] = 1000] = "SECOND";
})(TIME_IN_MILLISECONDS || (TIME_IN_MILLISECONDS = {}));
